setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php

```diff
@@ -52,46 +104,66 @@ public function getList($path)
             \RecursiveIteratorIterator::SELF_FIRST
         );
 
+        $classes = $this->extract($recursiveIterator);
+        if (!$isGeneration) {
+            $this->fileResults[$realPath] = $classes;
+        }
+        return $classes;
+    }
+
+    /**
+     * Extracts all the classes from the recursive iterator
+     *
+     * @param \RecursiveIteratorIterator $recursiveIterator
+     * @return array
+     */
+
+    private function extract(\RecursiveIteratorIterator $recursiveIterator)
+    {
         $classes = [];
         foreach ($recursiveIterator as $fileItem) {
             /** @var $fileItem \SplFileInfo */
             if ($fileItem->isDir() || $fileItem->getExtension() !== 'php') {
                 continue;
             }
+            $fileItemPath = $fileItem->getRealPath();
             foreach ($this->excludePatterns as $excludePatterns) {
-                if ($this->isExclude($fileItem, $excludePatterns)) {
+                if ($this->isExclude($fileItemPath, $excludePatterns)) {
                     continue 2;
                 }
             }
-            $fileScanner = new FileScanner($fileItem->getRealPath());
+            $fileScanner = new FileClassScanner($fileItemPath);
             $classNames = $fileScanner->getClassNames();
-            foreach ($classNames as $className) {
-                if (empty($className)) {
-                    continue;
-                }
-                if (!class_exists($className)) {
-                    require_once $fileItem->getRealPath();
-                }
-                $classes[] = $className;
-            }
+            $this->includeClasses($classNames, $fileItemPath);
+            $classes = array_merge($classes, $classNames);
         }
         return $classes;
     }
 
+    protected function includeClasses(array $classNames, $fileItemPath)
+    {
+        foreach ($classNames as $className) {
+            if (!class_exists($className)) {
+                require_once $fileItemPath;
```

Comment:
Presumably because of the possibility of an incorrect class definition in one file affecting another operation.  This is is how it was done in the original code.  https://github.com/magento/magento2/blob/364635691c8baa5cc30d58c1c1879fc309f5b048/setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php